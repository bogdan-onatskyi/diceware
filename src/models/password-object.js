import {observable, computed, action} from 'mobx';

import Word from './word-object';

class Password {
    @observable usedWords;
    @observable maxWords;

    @observable wordArray;
    @observable isPassboxOpened;
    // @observable caps;

    constructor() {
        this.usedWords = 5;
        this.maxWords = 8;
        this.wordArray = [];

        this.caps = 0;
        this.separators = ['-', '_', ':', '.', '', ' '];
        this.separators.forEach((s, i) => {
            this.caps += Math.pow(2, i);
        });

        this.isCAPS = (i) => this.caps & Math.pow(2, i);
        this.toggleCAPS = (i) => {
            this.caps = this.caps ^ Math.pow(2, i);
        };

        this.isPassboxOpened = true;
        this.toggleIsPassboxOpened = () => {
            this.isPassboxOpened = !this.isPassboxOpened;
        };

        this.handleUsedWords = [];
        for (let id = 1; id <= this.maxWords; id++)
            this.handleUsedWords [id - 1] = () => {
                this.wordArray.splice(id);
                for (let i = this.usedWords; i < id; i++)
                    this.wordArray.push(new Word(i));
                this.usedWords = id;
            };

        this.handleResetAllWords = () => {
            for (let i = 0; i < this.usedWords; i++)
                this.wordArray[i] = new Word(i);
        };

        this.separatePassword = (separator, i = -1) => {
            const isUpperCase = (i === -1) ? true : this.isCAPS(i);
            const s = this.password.join(separator);
            return isUpperCase ? s.toUpperCase() : s.toLowerCase();
        };

    }

    // @action('isCAPS')
    // isCAPS(i) {
    //     return this.caps & Math.pow(2, i);
    // };

    init(usedWords = 5, maxWords = 8) {
        this.usedWords = usedWords;
        this.maxWords = maxWords;
        this.handleResetAllWords();
    }

    @computed
    get password() {
        const pass = [];
        this.wordArray.forEach((wordObject) => pass.push(wordObject.word));
        return pass;
    }

    // @computed
    // get caps() {
    //     return this.caps;
    // }

    // @action('Caps state changed')
    // isCAPS(i) {
    //     return this._isCAPS(i);
    // }
}

export {Password};

const passwordObject = new Password();
export default passwordObject;