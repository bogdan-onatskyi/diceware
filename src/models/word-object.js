import {observable, computed, action} from 'mobx';

import wordList from './word-list';

class Word {
    @observable _index;

    constructor(wordBoxId) {
        this._id = wordBoxId;
        this._index = Word.getRandomIndex();
        this.handleClick = (e) => {
            const handler = {
                "prev2": () => this.ResetWord(this.prev2index),
                "prev1": () => this.ResetWord(this.prev1index),
                "next1": () => this.ResetWord(this.next1index),
                "next2": () => this.ResetWord(this.next2index),
                "minus": () => this.MinusWord(),
                "reset-word": () => this.ResetWord(),
                "plus": () => this.PlusWord(),
                "code": () => this.ResetWord()
            };
            for (let prop in handler) {
                if (e.target.className.includes(prop)) {
                    // e.preventDefault();
                    handler[prop]();
                    return;
                }
            }
        };
        this.handleWheel = (e) => {
            e.preventDefault();
            if (!e.target.className.includes("wv__word")) return;
            e.deltaY > 0 ? this.PlusWord() : this.MinusWord();
        };
    }

    get id() {
        return this._id;
    }

    @computed
    get index() {
        return this._index;
    }

    set index(value) {
        this._index = (value >= 0 && value < wordList.length) ? value : 0;
    };

    @computed
    get code() {
        return Word.indexToCode(this._index);
    }

    @computed
    get prev2word() {
        return Word.getWord(Word.getPrevIndex(Word.getPrevIndex(this._index)));
    }

    @computed
    get prev1word() {
        return Word.getWord(Word.getPrevIndex(this._index));
    }

    @computed
    get prev2index() {
        return Word.getPrevIndex(Word.getPrevIndex(this._index));
    }

    @computed
    get prev1index() {
        return Word.getPrevIndex(this._index);
    }

    @computed
    get word() {
        return Word.getWord(this._index);
    }

    @computed
    get next1index() {
        return Word.getNextIndex(this._index);
    }

    @computed
    get next2index() {
        return Word.getNextIndex(Word.getNextIndex(this._index));
    }

    @computed
    get next1word() {
        return Word.getWord(Word.getNextIndex(this._index));
    }

    @computed
    get next2word() {
        return Word.getWord(Word.getNextIndex(Word.getNextIndex(this._index)));
    }

    @action('Previous word was selected')
    MinusWord() {
        this.index = Word.getPrevIndex(this.index);
    }

    @action('New random word was generated')
    ResetWord(index = 0) {
        this.index = index !== 0 ? index : Word.getRandomIndex();
    };

    @action('Next word was selected')
    PlusWord() {
        this.index = Word.getNextIndex(this.index);
    }

    static indexToCode(index) {
        let localIndex = index;
        let code = '';
        let pow6, n;

        for (let i = 0; i < 4; i++) {
            pow6 = Math.pow(6, 4 - i);
            n = Math.floor((localIndex) / pow6);
            localIndex -= n * pow6;
            code += n += 1;
        }
        code += ++localIndex;

        return code;
    }

    static codeToIndex(code) {
        const codeArray = [] = code.split('');

        let index = codeArray[4] - 1;

        for (let i = 0; i < 4; i++) {
            index += (codeArray[i] - 1) * Math.pow(6, 4 - i);
        }

        return index;
    }

    static getRandomInt(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }

    static getRandomIndex() {
        return Word.codeToIndex(Word.getRandomCode());
    }

    static getRandomCode() {
        let code = '';
        for (let i = 5; i > 0; i--) code += Word.getRandomInt(1, 6);
        return code;
    }

    static getPrevIndex(index) {
        return index > 0 ? index - 1 : wordList.length - 1;
    }

    static getNextIndex(index) {
        return index < wordList.length - 1 ? index + 1 : 0;
    }

    static getWord(index) {
        return wordList[(index >= 0 && index < wordList.length) ? index : 0].toUpperCase();
    }
}

export default Word;