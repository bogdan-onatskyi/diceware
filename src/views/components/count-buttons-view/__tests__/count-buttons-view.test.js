import React from 'react';

import CountButtonsView from '../count-buttons-view';
import {Password} from '../../../../models/password-object';

describe("CountButtonsView component", () => {
    const passwordObject = new Password();

    const setup = ({
                       usedWords = passwordObject.usedWords,
                       maxWords = passwordObject.maxWords,
                       handleUsedWords = passwordObject.handleUsedWords,
                       handleResetAllWords = passwordObject.handleResetAllWords
                   }) => {

        const wrapper = shallow(<CountButtonsView usedWords={usedWords}
                                                  maxWords={maxWords}
                                                  handleUsedWords={handleUsedWords}
                                                  handleResetAllWords={handleResetAllWords}/>
        );
        return {wrapper};
    };

    it("should match saved snapshot", () => {
        const {wrapper} = setup({});
        expect(wrapper).toMatchSnapshot();
    });

    it("should set passwordObject.UsedWords", () => {
        const handleUsedWords = spyOn(passwordObject, 'handleUsedWords').and.callThrough();
        const {wrapper} = setup({handleUsedWords: handleUsedWords});

        wrapper.find({type: "count"}).forEach((button, i) => {
            button.simulate('click');
            expect(passwordObject.usedWords).toEqual(i + 1);
            expect(handleUsedWords).toHaveBeenCalledWith(i + 1);
        });
    });

    it("should reset all words in passwordObject.wordArray", () => {
        const handleResetAllWords = spyOn(passwordObject, 'handleResetAllWords').and.callThrough();
        const {wrapper} = setup({handleResetAllWords: handleResetAllWords});

        const wordArray = passwordObject.wordArray.map((wordObject) => wordObject.word);

        wrapper.find({type: "reset-all-words"}).simulate('click');
        expect(handleResetAllWords).toHaveBeenCalled();

        wordArray.forEach((word, i) => {
            expect(passwordObject.wordArray[i]).not.toEqual(word);
        });
    });

    it("should have 4 props", () => {
        const propsStr = 'usedWords maxWords handleUsedWords handleResetAllWords';
        const props = [];
        for (let prop in CountButtonsView.propTypes) props.push(prop);
        expect(props.join(' ') === propsStr).toBe(true);
    });
});