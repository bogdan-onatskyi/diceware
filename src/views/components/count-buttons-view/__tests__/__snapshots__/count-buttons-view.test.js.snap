// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountButtonsView component should match saved snapshot 1`] = `
<div
  className="buttons-view"
>
  <Button
    data-tip="Выбрать количество слов в пароле: 1"
    disabled={false}
    key="count-button_1"
    onClick={[Function]}
    type="count"
  >
    1
  </Button>
  <Button
    data-tip="Выбрать количество слов в пароле: 2"
    disabled={false}
    key="count-button_2"
    onClick={[Function]}
    type="count"
  >
    2
  </Button>
  <Button
    data-tip="Выбрать количество слов в пароле: 3"
    disabled={false}
    key="count-button_3"
    onClick={[Function]}
    type="count"
  >
    3
  </Button>
  <Button
    data-tip="Выбрать количество слов в пароле: 4"
    disabled={false}
    key="count-button_4"
    onClick={[Function]}
    type="count"
  >
    4
  </Button>
  <Button
    data-tip=""
    disabled={true}
    key="count-button_5"
    onClick={[Function]}
    type="count"
  >
    5
  </Button>
  <Button
    data-tip="Выбрать количество слов в пароле: 6"
    disabled={false}
    key="count-button_6"
    onClick={[Function]}
    type="count"
  >
    6
  </Button>
  <Button
    data-tip="Выбрать количество слов в пароле: 7"
    disabled={false}
    key="count-button_7"
    onClick={[Function]}
    type="count"
  >
    7
  </Button>
  <Button
    data-tip="Выбрать количество слов в пароле: 8"
    disabled={false}
    key="count-button_8"
    onClick={[Function]}
    type="count"
  >
    8
  </Button>
  <div>
    <Button
      data-tip="Изменить все слова в пароле"
      onClick={[Function]}
      type="reset-all-words"
    >
      Изменить все
    </Button>
  </div>
  <ToolTip />
</div>
`;
